File: go/2022/day07/day07.go
package main

import (
	"aoc/cast"
	"aoc/utils"
	"flag"
	"fmt"
	"math"
	"os"
	"strings"
)

var inputFile = flag.String("inputFile", "input.txt", "Relative file path to use as input")

func main() {
	flag.Parse()
	bytes, err := os.ReadFile(*inputFile)
	if err != nil {
		panic("Unable to read the input file")
	}

	input := string(bytes)

	fmt.Printf("AoC 2022, Day7, Part1 solution is: %d \n", part1(input))
	fmt.Printf("AoC 2022, Day7, Part2 solution is: %d \n", part2(input))
}

func part1(input string) int {
	root := parseInput(input)
	return sumDirsUnder100000(root)
}

func part2(input string) int {
	root := parseInput(input)
	totalSpaceAvailable := 70000000
	spaceNeeded := 30000000

	directoryMinSize := spaceNeeded - (totalSpaceAvailable - root.totalSize)
	return findSmallestDirToDelete(root, directoryMinSize)
}

func sumDirsUnder100000(itr *dir) int {
	SizeLimit := 100000

	sum := 0
	if itr.totalSize <= SizeLimit {
		sum += itr.totalSize
	}
	for _, child := range itr.childDirs {
		sum += sumDirsUnder100000(child)
	}
	return sum
}

func findSmallestDirToDelete(currentDir *dir, directoryMinSize int) int {
	smallest := math.MaxInt64
	if currentDir.totalSize >= directoryMinSize {
		smallest = utils.MinInt(smallest, currentDir.totalSize)
	}

	for _, childDirs := range currentDir.childDirs {
		smallest = utils.MinInt(smallest, findSmallestDirToDelete(childDirs, directoryMinSize))
	}

	return smallest
}

type dir struct {
	name      string
	parentDir *dir
	files     map[string]int  // map contains the name of the file and size
	childDirs map[string]*dir // map containing the directory name an a pointer to actual dir
	totalSize int
}

func parseInput(input string) *dir {
	root := &dir{
		name:      "root",
		childDirs: map[string]*dir{},
	}

	currentDir := root
	commands := strings.Split(input, "\n")

	index := 0

	for index < len(commands) {
		cmd := commands[index]
		switch cmd[0:1] {
		case "$":
			if cmd == "$ ls" {
				index++
			} else {
				changeDir := strings.Split(cmd, "cd ")[1]
				changeDir = strings.TrimSpace(changeDir)
				// if the change dir is .. => we need to go up one level
				if changeDir == ".." {
					currentDir = currentDir.parentDir
				} else {
					// if change dir doesn't exist, we need to create it
					if _, ok := currentDir.childDirs[changeDir]; !ok {
						currentDir.childDirs[changeDir] = &dir{
							name:      changeDir,
							parentDir: currentDir,
							childDirs: map[string]*dir{},
							files:     map[string]int{},
						}
					}
					currentDir = currentDir.childDirs[changeDir]
				}
				index++
			}
		default:
			//listening a dir content and add them to the current dir in case they don't exist
			if strings.HasPrefix(cmd, "dir") {
				childDirName := cmd[4:]
				if _, ok := currentDir.childDirs[childDirName]; !ok {
					currentDir.childDirs[childDirName] = &dir{
						name:      childDirName,
						parentDir: currentDir,
						childDirs: map[string]*dir{},
						files:     map[string]int{},
					}
				}
			} else {
				// file name
				parts := strings.Split(cmd, " ")
				currentDir.files[parts[1]] = cast.ToInt(parts[0])
			}
			index++
		}
	}

	populateFileSize(root)
	return root
}

func populateFileSize(currentDir *dir) int {
	totalSize := 0
	for _, childDir := range currentDir.childDirs {
		totalSize += populateFileSize(childDir)
	}

	for _, size := range currentDir.files {
		totalSize += size
	}

	currentDir.totalSize = totalSize
	return totalSize
}

File: python/2022/day07/day07.py
class TreeNode:
    def __init__(self, is_dir: bool, name: str, size: int = None) -> None:
        self.is_dir = is_dir
        self.name = name
        self.size = size
        self.children = []
        self.parent = None

    def add_child(self, child):
        child.parent = self
        self.children.append(child)

    def get_size(self):
        if self.is_dir:
            total_size = 0
            for child in self.children:
                total_size += child.get_size()
            return total_size
        else:
            return self.size

    def print_children(self, level):
        if self.is_dir:
            print("--" * level + self.name + " (total=" + str(self.get_size()) + ")")
        else:
            print("--" * level + self.name + " (file=" + str(self.get_size()) + ")")
        if len(self.children) > 0:
            for child in self.children:
                child.print_children(level + 1)

    def find_subdirectories_part1(self):
        dir_sizes = 0
        if self.is_dir:
            for child in self.children:
                if child.is_dir and child.get_size() <= 100000:
                    dir_sizes += child.get_size() + child.find_subdirectories_part1()
                else:
                    dir_sizes += child.find_subdirectories_part1()
        return dir_sizes

    def find_subdirectories_part2(self, min_size):
        dir_sizes = []
        if self.is_dir:
            for child in self.children:
                if child.is_dir and child.get_size() >= min_size:
                    dir_sizes += [child.get_size()] + child.find_subdirectories_part2(
                        min_size
                    )
                else:
                    dir_sizes += child.find_subdirectories_part2(min_size)
        return dir_sizes


class Tree:
    def __init__(self) -> None:
        self.root = TreeNode(is_dir=Tree, name="root")
        self.current = self.root

    def reset_to_root(self):
        self.current = self.root

    def go_up_one_level(self):
        self.current = self.current.parent

    def go_to_child(self, name):
        self.current = list(
            filter(lambda child: child.name == name, self.current.children)
        )[0]

    def add_new_child(self, child):
        self.current.add_child(child)


def create_tree(input: str) -> Tree:
    tree = Tree()

    lines = parse_input(input)

    while len(lines) > 0:
        line = lines.pop(0)
        if line == "$ cd /":
            tree.reset_to_root()
        elif line == "$ ls":
            while len(lines) > 0 and "$" not in lines[0]:
                line = lines.pop(0)
                size, name = line.split(" ")
                if size.isdigit():
                    new_node = TreeNode(is_dir=False, name=name, size=int(size))
                else:
                    new_node = TreeNode(is_dir=True, name=name)
                tree.add_new_child(new_node)
        elif line == "$ cd ..":
            tree.go_up_one_level()
        elif "$ cd" in line:
            _, _, name = line.split(" ")
            tree.go_to_child(name)

    return tree


def part1(input: str) -> int:
    tree = create_tree(input)
    return tree.root.find_subdirectories_part1()


def part2(input: str) -> int:
    tree = create_tree(input)
    
    total_space = 70000000
    space_needed = 30000000
    current_empty_space = total_space - tree.root.get_size()
    possible_dirs = tree.root.find_subdirectories_part2(space_needed - current_empty_space)
    return min(possible_dirs)


def parse_input(input: str) -> list[str]:
    with open(input) as fn:
        raw_data = fn.read()
        parts = raw_data.split("\n")
    return parts


def main():
    puzzle_input_path = "2022/day07/input.txt"
    print(f"AoC2022 Day7, Part1 solution is: {part1(input=puzzle_input_path)}")
    print(f"AoC2022 Day7, Part2 solution is: {part2(input=puzzle_input_path)}")


if __name__ == "__main__":
    main()
